File: selectionSort.js
"use strict";

// GIVEN
// E A D H B

// first pass looks for the minimal value, and swaps it with the value at the front of the list
// A E D H B

// next pass finds the minimal value after the first element (which is now in place) and swaps it.
// so from e, d, h, b..b is smallest and swap with the next value, which is e.
// A B D H E

// move on, and do another pass , D, H, E.
// D is smallest so everything stays
// A B D H E

// move on, and do another pass. H, E.
// E is smaller is swap H and E.
// A B D E H

var dataArray = [8,3,1,6,2];

function swap(array, index1, index2) {
  var temp = array[index1];
  array[index1] = array[index2];
  array[index2] = temp;
}

// given a certain pass
// we get the minimum value and swap it
function swapSmallerValuesToFront(data, outerIndex, lastIndex) {
  for (var innerIndex = outerIndex + 1; innerIndex <= lastIndex; innerIndex++) {
      console.log("");
      console.log("inner index: " + innerIndex + ", data[inner]: " + data[innerIndex]);
      console.log("data[outerIndex]: " + data[outerIndex]);

      if (data[innerIndex] < data[outerIndex]) {
          console.log(" We need to swap because current element is smaller");
          swap(data, innerIndex, outerIndex);
          console.log("---> data display: " + data);
      }
  }
}

function selectionSort(intArray) {

    var data = intArray.slice(0);
    var indexOfMinValue, temp;
    var lastIndex = data.length-1;

    // we always do length - 1 passes
    for (var pass = 0; pass <= lastIndex - 1; pass++) {
      console.log("---------------> Outer Index: " + pass + " <-----------");
      swapSmallerValuesToFront(data, pass, lastIndex);
    }

    return data;
}

let sortedArray = selectionSort(dataArray);
console.log(sortedArray);

File: selectionSort.cpp
#include <iostream>

using namespace std;

void swap(int * arrPtr, int index1, int index2) {
    int temp = arrPtr[index1];
    arrPtr[index1] = arrPtr[index2];
    arrPtr[index2] = temp;
}

void swapSmallerValuesToFront(int arr [],
                              int outerIndex,
                              int lastIndex) {
    
    for (int innerIndex = outerIndex + 1; innerIndex <= lastIndex; innerIndex++) {
        // the value at outerIndex may always change because
        // it may be switched with previous elements that are smaller
        // than it. That's why we must access the outerIndex value
        // every time here.
        if (arr[innerIndex] < arr[outerIndex]) {
            cout << " We need to swap because current element is smaller" << endl;
            swap(arr, innerIndex, outerIndex);
        }
    }
}


void selectionSort(int arr [], int size) {
    cout << "size: " << size << endl;
    int lastIndex = size - 1;
    for (int pass = 0; pass <= lastIndex - 1; pass++) {
        cout << "----> Outer Index: " << pass << endl;
        swapSmallerValuesToFront(arr, pass, lastIndex);
    }
}
                   
int main() {
    
    int intArray [] = {13,61,44,88,1,34,54,12,22,2,4,6,3,5,7};
    int arrSize = sizeof(intArray)/sizeof(int);
    
    selectionSort(intArray, arrSize);
    
    // display the results
    for ( int i = 0; i < arrSize; i++) {
        cout << intArray[i] << endl;
    }
    
    return 0;
}
